firestore/src/main/aggregate_query_main.cc

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "firestore/src/main/aggregate_query_main.h"
+
+#include "Firestore/core/src/api/aggregate_query.h"
+#include "Firestore/core/src/api/query_core.h"
+#include "firestore/src/include/firebase/firestore/aggregate_query.h"
+#include "firestore/src/include/firebase/firestore/aggregate_query_snapshot.h"
+#include "firestore/src/main/aggregate_query_snapshot_main.h"
+#include "firestore/src/main/listener_main.h"
+#include "firestore/src/main/promise_factory_main.h"
+#include "firestore/src/main/util_main.h"
+
+namespace firebase {
+namespace firestore {
+
+AggregateQueryInternal::AggregateQueryInternal(
+    api::AggregateQuery&& aggregate_query)
+    : aggregate_query_{std::move(aggregate_query)},
+      promise_factory_{PromiseFactory<AsyncApis>::Create(this)} {}
+
+FirestoreInternal* AggregateQueryInternal::firestore_internal() {
+  return GetFirestoreInternal(&aggregate_query_.query());
+}
+
+const FirestoreInternal* AggregateQueryInternal::firestore_internal() const {
+  return GetFirestoreInternal(&aggregate_query_.query());
+}
+
+Query AggregateQueryInternal::query() {
+  return MakePublic(api::Query(aggregate_query_.query()));
+}
+
+Future<AggregateQuerySnapshot> AggregateQueryInternal::Get(
+    AggregateSource source) {
+  auto promise =
+      promise_factory_.CreatePromise<AggregateQuerySnapshot>(AsyncApis::kGet);
+  // TODO(tomandersen) Is there a better way to pass `aggregate_query_` than
+  // through `this`
```

Comment:
RE: "Is there a better way to pass `aggregate_query_` than through `this`"...

Yes, and I believe you _should_ do it differently because `this` may become invalid before the callback gets invoked.

The way I'd do it is create a local variable, `auto aggregate_query = aggregate_query_` then pass it by value into the lambda using the lambda's capture clause, instead of passing `this` in the lambda's capture clause.